### PR TITLE
Revert "nginx: use /etc/nginx/nginx.conf enabling conf.d/"

### PR DESCRIPTION
**Pull request by peterwillcn**
This reverts commit 2401fd6db5e582d1b1442e4832222a3fee50937a.

Nginx is nginx does not need nginx-ssl-util，I think the commit should revert!

### CODE REVIEW

**Comment by hnyman** [comment]
cc @Ansuel  @peter-stadler

**Comment by peter-stadler** [comment]
The nginx-ssl-util would be for setting up the default server on the LAN with a self-signed certificate, the previous version did that with a shell script using another utility (px5g or openssl-util) ...
openssl-util is bigger and px5g is linked against mbedtls as default (while nginx-ssl depends on libopenssl). So, I made a tool that is basically px5g linked against libopenssl and including the shell script.
It is checking on startup if the default server on LAN has a valid certificate and recreates it if needed. I think it is better to do that on startup than on the first install as it could expire. I would have also a version that does the same in the script /etc/init.d/nginx, but that is slower (adding about 400ms on startup on my system)

**Comment by peterwillcn** [comment]
@peter-stadler 
I think nginx software package and other packages do not have dependencies, which will cause the system to be complex, the more complex the system is more prone to problems, you can modify this modification according to their own needs to modify the configuration file can be solved, also not common. :-)

**Comment by peter-stadler** [comment]
I am with you that it should be not complex. Right now the tool is quite simple and uses some of the dependencies that Nginx has `+libstdcpp` (I am looking to replace that with `uclibc++`).

It can do the following things:
* `nginx-util add_ssl FQDN`: Create a self-signed certificate for a FQDN and add the corresponding directives to `/etc/nginx/conf.d/FQDN.conf` if that file exists and has FQDN in the server_name directive. If cron is available it sets also a job to renew the certificate.
* `nginx-util del_ssl FQDN`: Remove that again.
* `nginx-util init_lan`: Get the IP addresses of loopback and lan and do basically `nginx-util add_ssl _lan` that is the default server for the LAN.
* `nginx-util get_env`: Exports the internal constants as environment variables.

We are thinking to setup some [uci](https://openwrt.org/docs/techref/uci) configuration for Nginx, and maybe to use this tool to load the configuration.

And I am really sorry, I do not understand your second sentence.

**Comment by peter-stadler** [comment]
The new configuration file `/etc/nginx/nginx.conf` is meant to be a minimal default configuration that can be extended by putting files in `/etc/nginx/conf.d/`. We had the problem that other packages would modify the main config, e.g.:
  * `luci-nginx/luci-ssl-nginx` is replacing the file and enables also `/etc/nginx/conf.d/`
  * `ariang-nginx` would include locations below the `server_name`
  * `etesync-server` is a package that I want to add and that would have to install its own locations

Beside that, there was a default server in the main config, that listened on all IPs; so you cannot install another default server, without modifying this fail further. The new version restricts the default server (now in `/etc/nginx/conf.d/_lan.conf`) to IP addresses of the LAN (and loopback), so you can add `listen 80 default_server;` to another server in `/etc/nginx/conf.d/`.

This is inspired by [Debian](https://wiki.debian.org/Nginx/DirectoryStructure) who use some more folders, for comparison (without the upstream files):
debian: `./conf.d/*.conf` -----------------------------------------> openwrt: `./conf.d/*.conf`
debian: `./sites-available/*` ----------------------------------> openwrt: `./conf.d/*.conf.disabled`
debian: `./sites-enabled/*` --------------------------------------> openwrt: `./conf.d/*.conf`
debian: `./snippets/*.conf` --------------------------------------> openwrt: `./conf.d/*.something.else`
debian: `./apps.d/*.conf` or `./default.d/*.conf` ----> openwrt: `./conf.d/*.locations`

Currently we store also certificates and their keys in `./conf.d/*.{crt,key}` ...

**Comment by neheb** [comment]
> (I am looking to replace that with `uclibc++`).

Why? It only supports C++98.

**Comment by peter-stadler** [comment]
@neheb mostly because it is the standard value for the preferred C++ library (in menuconfig), is it not on the way to support more? But, if I change that then only on the long run …

**Comment by neheb** [comment]
uClibc++ is also broken in subtle ways, in addition to being dead upstream.

I'd rather replace it with something like libcxx( which is in the tree). libcxx doesn't build on several platforms though.

It's also bigger than uClibc++.

**Comment by peter-stadler** [comment]
I see, so I will stick with libstdcpp for now and look into libcxx later :-)

File: net/nginx/Makefile [hunk 1]
```diff
@@ -96,16 +95,13 @@ endef
 
 define Package/nginx
   $(Package/nginx/default)
-  DEPENDS += +!NGINX_SSL:nginx-util +NGINX_SSL&&NGINX_PCRE:nginx-ssl-util \
```
Comment (peterwillcn) [inline net/nginx/Makefile]: Nginx should not rely on nginx-util, which is optional

**Comment by peter-stadler** [comment]
@peterwillcn I think we should look at this in two steps: 
     (A) What we want and 
     (B) How we achieve it:

(A) 
1. Should we use one big `/etc/nginx/nginx.conf` file including only nginx.org parts?
 (yes) How should other packages be installed? In the revert LuCI replaces the file, ariang adds an include.
 (no) How should the folder structure look like? Debian uses five folders mostly with `*.conf` files. I propose to use only the `conf.d` folder, packages can install `*.locations` there and they would be available on LAN as default; the admin could install servers in `FQDN.conf` files (I would look into making that available through `uci` and maybe LuCI, too).
2. Should we have a default server? In the revert there is one that listens on all IPs. I propose to use one listening on local IPs only (that includes all locations), then an admin can install another one for the WAN, which exposes only selected locations (=apps).
(yes) How should it be setup?
(no) How do we install apps, e.g. LuCI should be available after installation?
3. I think we should be able to create self-signed certificates for nginx-ssl: It should be encryption on by default, but we cannot assume that the admin has a valid domain and installs SSL on it. So, it is the best if we install a self-signed certificate at least for the default server (on the LAN). This is also what we do already for other servers with SSL and LuCI. 
Should we make this behaviour configurable?

(B)
1. Setting up `/etc/nginx/nginx.conf` and the folder structure is straight forward: change it and adopt `/etc/init.d/nginx` as well as the wiki to it.
If we use `uci` to setup some configuration we can read it with sh in the init file or adopt `nginx-util` to do it (This is the road to expose some of the Nginx configuration to LuCI, too).
2. If we want the default server for the LAN, we need to get its IPs somehow. 
The `nginx-util` gets them like 
`ifstatus lan | jsonfilter -e '@["ipv4-address","ipv6-address"].*.address'`
 but through the C interface (ubus-cxx wraps it to C++ syntax):
```
    auto lan_status = ubus::call("network.interface.lan", "status");

    for (auto ip : lan_status.filter("ipv4-address", "", "address")) {
        add_listen("",  static_cast<const char *>(blobmsg_data(ip)), "");
    }

    for (auto ip : lan_status.filter("ipv6-address", "", "address")) {
        add_listen("[", static_cast<const char *>(blobmsg_data(ip)), "]");
    }
```
3. For creating self-signed certificates, we have `px5g` using `mbedtls` (while `nginx-ssl` depends on `libopenssl`) and `openssl-util` that is larger (providing a lot more options). I think the best is to use something like `px5g` linked against `libopenssl`.
The `nginx-ssl-util` is basically `px5g-openssl` (using better encryption than rsa:512) plus setting up the corresponding configuration (what we would end up doing in a bash script).
Which option should we use? I prefer the `nginx-util` as I created it to fit this purpose.
If we make it configurable, I would move the src/nginx-util to nginx and adopt the Makefile of nginx, else it is a lot more of `DEPENDS += +CONFIG_xxx&&CONFIG_yyy:zzz` or are you seeing another way?
(IMHO it is not an option to leave the certificates away altogether, this would be a regression to the previous state in terms of security)

**Comment by peterwillcn** [comment]
A)
1,  yes；nginx-util is optional and not required for nginx.
The nginx-util function is to assist in the creation of sites-available (template files or backup configurations such as _lan.conf etc ...). Soft link to the sites-enabled and ssl (acme) folder and in / etc / nginx / nginx. conf added
include /etc/nginx/conf.d/*,conf; (general configuration such as blockips.conf)
include /etc/nginx/sites-enabled/*.conf; (on-line configuration, ln -s /etc/nginx/sites-available/xx.conf.)
include /etc/nginx/stream.d/*.conf; (stream forwarding configuration file)
2, lan default server, and port forward to wan
3, Use the SSL certificate generated by acme and place it in / etc / nginx / ssl. Modify the server configuration file in sites-enabled. You can use luci to modify the configuration file.

**Comment by peter-stadler** [comment]
I was thinking about your proposal, I suppose, you consider it obsolete as you edited it away? (Else: it would be achievable with quite some work for most of it, but we cannot use acme for non-public domains ...

I would work on a modification that you can configure the custom settings away. Would that be in your interest? And if so, should we install a `/etc/nginx/nginx.conf` in that case at all?

**Comment by peterwillcn** [comment]
@peter-stadler 
If there are special monitoring and 6-layer forwarding, docker can be used for isolation, so as to keep the original ecology of nginx design as far as possible, which is also the guiding principle for why openwrt USES patch.

**Comment by peter-stadler** [comment]
Now I can understand it better, thank you for extending the explanation :-) Although I cannot follow all of it, I completely agree with you for making it more flexible.

One clarification that I want to emphasize: The `nginx-util` is there for setting up some configuration files used by Nginx (currently: making listen directives with local IP addresses available, creating self-signed certificates and adding SSL directives if needed). It is no daemon and it does not interfere with the `nginx` process. There was always the `/etc/init.d/nginx` for controlling the run of it (I changed this file, too, but I suppose you agree with that beside the calls to `nginx-util`.)

I wanted to go one step in the direction to make it more flexible for users and package devs:
* The main configuration file includes all `/etc/nginx/conf.d/*.conf` files. So, there is a way to install further configuration without touching the main configuration file. And if the main `/etc/nginx/nginx.conf` is not modified it will be updated on package updates.
I assume that you are not opposing this change, since previously, the [official example ](https://hg.nginx.org/nginx/file/tip/conf/nginx.conf) was used. As it is not including any other files, it had to be changed (and then it is not updated anymore). This was done to make other apps (e.g. LuCI, ariang and in future etesync-server) available, so you cannot rely on which file is installed.
Although, now it is possible to leave the main configuration file untouched, we cannot prevent users or even packages to modify it. Right now this is needed to use the `stream` directive, for example.
I am open to adjust the `/etc/nginx/nginx.conf` if you have proposals, I just want it to have something like `include /etc/nginx/conf.d/*.conf` enabled by default.
If it is only about having a custom main configuration, we could also patch the official example (removing all comments and replacing the default server with the include would be quite similar).
* Two directives cannot be setup in the configuration files: the options `daemon off;` and `worker_processes "$(grep -c '^processor\s*:' /proc/cpuinfo)"`  passed by `/etc/init.d/nginx`. The first one is specifically for the init and was always there. The latter one, was previously changed in the main configuration file. We could go back to that (plus removing it on `prerm` to allow updates) if needed.
* There has always been a default server in the installed configuration, I wanted to keep that but move it to the file `/etc/nginx/conf.d/_lan.conf`. So, it can be modified without touching the main configuration (see above).
If LuCI is installed for Nginx with SSL, the default server was changed to use SSL with a self-signed certificate. I think it is a good idea to do that on the installation of Nginx (with or without LuCI). Furthermore, I did restrict the default server to the LAN. So, not all locations are automatically exposed to the WAN, when you make it available. Instead, the user can install another `/etc/nginx/conf.d/FQDN.conf` that should only include specific locations.
I am not sure if you are opposing this changes, I made both with security by default in mind.

Do you have suggestions how to make it more flexible? On my side, I am planning:
1. I want to fix one thing: There is an error message, when the `_lan.conf` file does not contain the default server. It is not doing any harm, but it is misleading as this is a reasonable thing to do if you want to remove the default server. (I think this is something you want to do.)
2. If you think it is a good idea, I would then propose to add a configuration option for allowing to build Nginx without the default server and nginx-util. (How should the main configuration look like then or should we omit it completely?)
3. In the long run, I would also like to make it possible to use `uci` (and then also LuCI) for setting up server parts of Nginx (and maybe some more configuration). This is done typically, when invoking `/etc/init.d/nginx [start|restart|reload]` and we can use `nginx-util` in more or less extend for helping with it.

From an architectural level: I want to integrate Nginx into the OpenWrt ecology leaving the Nginx ecology and the previous defaults intact as far as possible. 

From this point of view, we have:
* The new configuration is extensible like it was in some cases before (when `luci-nginx` was installed), but it is unified and the default server is included from another file now.
* The default server was changed by other packages to add their locations and install SSL to it. Now it is supposed to be changed only by admins and should provide a setup with security first using SSL if possible and exposing itself only to local IPs.
* There has always been a program to setup/start/stop/... Nginx. Before it was just the `/etc/init.d/nginx` script. Now this file is assisted by the `nginx-util` (first I wanted to include its functionality into the init script, but now I use C++ to gain performance).
* The `nginx-util` is used to setup the default server and it is available to create and use a self-signed certificate also for other servers. Right now it is modifying the corresponging `/etc/nginx/conf.d/FQDN.conf` files accordingly (In future it could create some of them completly from uci configuration).
* In the directory `/etc/nginx/conf.d/` can co-exist files created by packages (mostly `*.locations`), hand-crafted configuration files and files modified by `nginx-util`.
What have you in mind for designing the code on an architectural level?

Sorry, this post is getting (again) quite long, so I finish here with thanking you for your considerations.

**Comment by peter-stadler** [comment]
Made a [PR for nginx-util](https://github.com/openwrt/packages/pull/11284) addressing point 1. beside other things.

**Comment by neheb** [comment]
Closing this as nginx-util is under development.

**Comment by peter-stadler** [comment]
(The new draft is at #11456)

**Pull request closed without merging**